core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,588 @@
+      } else if (group.protocol != protocol || !group.supportsProtocols(subProtocols)) {
+        // if the new member does not support the group protocol, reject it
+        responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL.code))
+      } else if (group.not(PreparingRebalance) && group.has(memberId) && group.get(memberId).hasMetadata(metadata)) {
+        val member = group.get(memberId)
+        completeAndScheduleNextHeartbeatExpiration(group, member)
+        if (memberId == group.leaderId) {
+          if (group.is(Stable)) {
+            // if the leader sends JoinGroup while stable, we force a rebalance since this
+            // allows the leader to trigger rebalances for metadata changes
+            maybePrepareRebalance(group)
+            member.awaitingJoinCallback = responseCallback
+
+            if (group.is(PreparingRebalance))
+              rebalancePurgatory.checkAndComplete(ConsumerGroupKey(group.groupId))
+          } else {
+            // otherwise we are still awaiting sync from the leader. since it's possible
+            // that the leader missed the first JoinGroup response, we just resend it
+            responseCallback(JoinGroupResult(
+              members=group.currentMemberMetadata,
+              memberId=memberId,
+              generationId=group.generationId,
+              subProtocol=group.subProtocol,
+              leaderId=group.leaderId,
+              errorCode=Errors.NONE.code))
+          }
+        } else {
+          // for followers, just return group information and let them collect their assignment
+          responseCallback(JoinGroupResult(
+            members=Map.empty,
+            memberId=memberId,
+            generationId=group.generationId,
+            subProtocol=group.subProtocol,
+            leaderId=group.leaderId,
+            errorCode=Errors.NONE.code))
+        }
+      } else {
+        val member = if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          // if the member id is unknown, register the member to the group
+          val generatedMemberId = group.generateNextMemberId
+          val member = addMember(generatedMemberId, sessionTimeoutMs, metadata, subProtocols, group)
+          maybePrepareRebalance(group)
+          member
+        } else {
+          val member = group.get(memberId)
+          if (!member.hasMetadata(metadata) || subProtocols != member.subProtocols) {
+            // existing member changed its protocol metadata
+            updateMember(group, member, metadata, subProtocols)
+            maybePrepareRebalance(group)
+            member
+          } else {
+            // existing member rejoining a group due to restabilize
+            member
+          }
+        }
+
+        member.awaitingJoinCallback = responseCallback
+
+        if (group.is(PreparingRebalance))
+          rebalancePurgatory.checkAndComplete(ConsumerGroupKey(group.groupId))
+      }
+    }
+  }
+
+  def handleSyncGroup(groupId: String,
```

Comment:
In the current implementation we can return the sync response multiple times to members' requests as long as they have the right generation-id and are within the group. I think it does not have any side-effects for now but just want to clarify with you.